docs/CppInteroperability/ForwardVision.md

```diff
@@ -0,0 +1,290 @@
+
+If `StatefulObject` were written idiomatically in Swift, it would be defined as a `class` to make it a reference type. This is an example of how Swift defines clear patterns for naming, generic programming, value categories, error handling, and so on, which codebases are encouraged to use as standard practices. These well-defined programming patterns make using Swift APIs a cohesive experience, and C++ interoperability should stive to maintain this experience for Swift programmers using C++ APIs.
+
+To achieve that, the compiler should map C++ APIs to one of these specific Swift programming patterns. In cases where the most appropriate Swift pattern can be inferred by the Swift compiler, it should map the API automatically. Otherwise, Swift should ask programmers to annotate their C++ APIs to guide how they are imported. For example, Swift imports C++ types as structs with value semantics by default. Because `StatefulObject` cannot be copied, Swift cannot import it via the default approach. To be able to use `StatefulObject`, the user should annotate it as a reference type so that the compiler can import it as a Swift `class`. Information on how to import APIs, such as `StatefulObject`, cannot always be statically determined (for example, `StatefulObject` might have been a move-only type, a singleton, or RAII-style API). The Swift compiler should not import APIs like `StatefulObject` for which it does not have sufficent semantic information. It is not a goal to import every C++ API into Swift, especially without additional, required information to present the API in an idiomatic way that promotes a cohesive Swift expirence.
+
+Because of the difference in idioms between the two languages, and because of the safety concerns when exposing certain APIs to Swift, a C++ API might look quite different in Swift than it does in C++. It is a goal of C++ interoperability to provide a clear, well-defined mapping for whether and how APIs are imported into Swift. Users should be able to read the C++ interoperability documentation to have a good idea of how much of their API will be able to imported and what it will look like. Swift should also provide tools for inspecting what a C++ API will look like in Swift, and these tools should call out notable parts of the API that were not imported.
+
+Finally, Swift programmers have used C++ APIs since Swift 1.0 through an Objective-C or C bridging layer. It is important that programmers are able to *incrementally* remove these (potentially huge) bridging layers and start using their C++ APIs directly. To make this transition incremental, C++ interoperability must not change the way Swift imports any existing C or Objective-C APIs, even when C++ interoperability is enabled. 
+
+
+## The approach
+
```

Comment:
I think you've got the right ideas for what you want to say here, but it's missing a cohesive narrative.  Let me suggest how you can rework the first few paragraphs of this section.

```suggestion
The example above shows how the basic tools provided by C++ are often used in idiomatically different ways.  This poses a conundrum for importing C++ APIs automatically into Swift, and it's bigger than just reference vs. value semantics for classes.  A more fundamental place to see this is with memory management.

In Objective-C, it's fairly straightforward for ARC to ensure that data is valid when it's used.  Nearly all data in Objective-C is represented with either a fundamental type like `double` or `BOOL` or a managed object reference type like `NSString*`.  Fundamental values can be freely copied around, and managed object references can be conservatively retained to extend their lifetime.  It's rare to work with unmanaged C pointers in Objective-C, and even rarer to work with unmanaged pointer that have a dependency on a managed object (such as [`NSData`'s `-bytes` method](https://developer.apple.com/documentation/foundation/nsdata/1410616-bytes?language=objc)).  These latter cases can be treated as special cases and dealt with individually.

In contrast, it's very common for C++ APIs to work with unmanaged pointers, references, and views into other objects.  The lifetime rules for using these correctly are often varied, maybe even unique to an API.  As an example, consider three values: a value of type `std::vector<std::string>`, a reference returned from that vector's subscript operator, and an iterator returned from that vector's `begin()` method.  On a primitive level, these values look similar: they are all either pointers or class objects containing pointers.  But each has its own semantics and expected use, and these differences are not completely explicit in the source.  The vector is a value type that can be copied, but copies can be expensive, and iterators and references into the vector are valid only for a specific copy.  The subscript result is a mutable projection of a specific element, dependent on the vector for validity; but that element value can be copied out of the reference to get an independent value, and that is often how the subscript operator is used.  The iterator is also a projection, dependent on the vector for validity, but it must be used in conjunction with other iterators or with the vector itself in certain careful ways, and some operations will invalidate it completely.

So there is a conundrum where superficially similar language structures in C++ are used to express idiomatic patterns that are vastly different in their impact.  The only viable approach for addressing this conundrum is to pick off these patterns one at a time.  Swift's C++ importer will apply standard rules to sort C++ APIs into different idioms that it knows how to recognize.  The programmer will also be able to intervene by adding annotations to guide the importer.  If the standard rules don't allow the API to be imported, or if the API has an annotation that it doesn't actually satisfy the rules for, the importer will make the API unavailable, leaving enough breadcrumbs to tell the programmer what went wrong if they try to use it.

Consider how this applies to the `std::vector` example.  `std::vector<std::string>` maps over well as a Swift value type.  Its subscript operator can be imported as a Swift `subscript`, and the importer can take advantage of the fact that it returns a reference to allow elements to be efficiently borrowed.  And while C++ iterators in general pose serious lifetime safety problems in Swift, Swift can recognize the common `begin()`/`end()` idiom and import it as a safe Swift iterator that encapsulates the unsafety internally.
```